### DISCUSSION TITLE
Blazor Wasm support for DurableClientFactory

### DISCUSSION THREAD

**Discussion by Dresel** [Azure/azure-functions-durable-extension#1876] [Ideas]
Hi,

I was eager trying the DurableClientFactory functionality for calling entity functions in Blazor WASM and was a little bit disappointed that it does not work:

> The type or namespace name 'ApplicationPartAttributeAttribute' does not exist in the namespace 'Microsoft.AspNetCore.Mvc.ApplicationParts' (are you missing an assembly reference?)

Any chance that this will be supported in the future?

**Comment by cgillum** [comment]
I'm not aware of what work needs to be done for something like this. The error message you referenced doesn't seem related to Durable Functions.

@aaronpowell, given your experience with Blazor/Wasm and Durable Functions generally, do you know what we might need to do to get our .NET client to work in a Blazor Wasm app?

**Comment by aaronpowell** [comment]
Are we talking .NET 3.1 or .NET 5?

**Reply by Dresel** [reply]
Good catch, this error is only happening when targeting net5.0. Maybe related to this [issue](https://github.com/dotnet/aspnetcore/issues/27373).

These are the attributes mentioned by the compiler:

```csharp
//------------------------------------------------------------------------------
// <auto-generated>
//     This code was generated by a tool.
//     Runtime Version:4.0.30319.42000
//
//     Changes to this file may cause incorrect behavior and will be lost if
//     the code is regenerated.
// </auto-generated>
//------------------------------------------------------------------------------

using System;
using System.Reflection;

[assembly: Microsoft.AspNetCore.Mvc.ApplicationParts.ApplicationPartAttribute("Microsoft.AspNetCore.Mvc.WebApiCompatShim")]
[assembly: Microsoft.AspNetCore.Mvc.ApplicationParts.ApplicationPartAttribute("Microsoft.Azure.WebJobs.Extensions.DurableTask")]

// Generated by the MSBuild WriteCodeFragment class.


```

**Reply by aaronpowell** [reply]
That'd be my guess. Looking at the transitive dependency list of a Durable Functions app, there are several `*.Mvc.*` references that come in.

Having a dig around the code for Durable Functions, I don't think it'd be possible to fix without some radical changes to how Durable Functions is written. I'm also not sure that a `DurableClientFactory` is the right way to trigger entity functions in a client, as it would need to know about the transport model (Storage/Redis/etc.) that Durable Functions is using and be able to talk to it.

I'd be curious to understand more of what you're trying to do and see if there isn't another way to tackle the problem.

**Reply by aaronpowell** [reply]
Yeah, it looks like the `DurableClientFactory` needs access to the `IDurabilityProviderFactory` (which is Storage/Redis/etc., as specified in the Function app) to create a `DurableClient` for you to interact with, but then that uses `TaskHubClient` to use more stuff that I'm willing to try and dig through at this point in time 🤣

**Reply by Dresel** [reply]
> I'd be curious to understand more of what you're trying to do and see if there isn't another way to tackle the problem.

Well I want to execute entity (or azure in general) functions from blazor wasm. Combining static webassembly webapps with serverless computing is somehow appealing (if not to say sexy 😃).

That's the idea of this PR for the [IDurableClientFactory](https://github.com/Azure/azure-functions-durable-extension/pull/1125) if I'm not mistaken? I'm still new to azure functions so apologize if I mix something up.

**Reply by aaronpowell** [reply]
@cgillum can go into a lot more of the background to the PR and the motivation behind the API, but the way I look at it, it's still intended to be a server-side operation, not a client-side one (the use case illustrated is a controller action).

For me, enabling this within Blazor WASM would raise a pretty big red flag. Blazor is running in an untrusted security context, so you're having to expose a secured context to it to enable this functionality. And after all, Blazor is only able to talk HTTP(s) and WS(s), since it's running in a browser context (be that a web browser or hosted engine like Electron), unlike an aspnet controller.

Given that this is all doable via HTTP endpoints, aren't you already able to do all the serverless orchestration you want? or is the hurdle that the programming interface to do it via HTTP not adequate?

**Reply by Dresel** [reply]
That makes sense.

> Given that this is all doable via HTTP endpoints, aren't you already able to do all the serverless orchestration you want? or is the hurdle that the programming interface to do it via HTTP not adequate?

I lived under the assumptions that the durableclient is anyway calling http endpoints. But yes, this should work.

You are referring to [this](https://docs.microsoft.com/en-gb/azure/azure-functions/durable/durable-functions-http-api) right? Do you know existing clients based on this or does a swagger spec exist?

**Reply by Dresel** [reply]
Anyway thanks for your time and answers. Very much appreciated.
